sdk/keyvault/azure-security-keyvault-administration/test/ut/CMakeLists.txt

```diff
@@ -22,6 +22,8 @@ add_executable (
     settings_client_test.cpp
 )
 
+target_compile_definitions(azure-security-keyvault-administration-test PRIVATE _azure_BUILDING_TESTS)
```

Comment:
This seems tedious to have to specify for each file. Could we "detect" it in a cmake macro, with any file that is in `/test/` will have `_azure_BUILDING_TESTS`, anything in a `/samples/` directory will be `_azure_BUILDING_SAMPLES`, and the default would be `_azure_BUILDING_SDK` to cover the rest/default value.

Reply:
One potential place to put this logic could be in a macro we already call `create_per_service_target_build`, and this way, you don't have to modify all these 100+ files (assuming you could find a way to get the file path to detect the test/samples directory).

Reply:
I am not sure if `create_per_service_target_build` macro is used universally, and I don't think it is the best idea to hide the fact that we set the macro - it is easier to realize everything is being set as expected by seeing it. Each test has multiple things that it sets, why this one should be an exception.

Reply:
> I am not sure if `create_per_service_target_build` macro is used universally,

Shouldn't it be? I thought we'd want to have consistency in a lot of this boilerplate.

If we have a lot of ad-hoc behavior it:
1) Makes things harder to understand and follow, and therefore error-prone.
2) We cannot code-gen it, even in the future

Without this, new files/services might easily miss this and skew the results.

Reply:
74 files have it.
All attestation `CMakeLists.txt` have it.

It is only the "root" CMakeLists.txt for each SDK, the sample "roots" and the "vendor" files that don't have it. And these files are pretty barebones and just route to the other specific CMakeLists.txt with the real content in them.

Like:
> sdk\attestation\CMakeLists.txt
> sdk\attestation\azure-security-attestation\samples\CMakeLists.txt

Out of the  43 files with the missing line (out of Total non-vendor files: 116), only a handful seem to be missing it. 8 are root SDK, 6 are root sample, 25 are core, which just leaves these 4, which are all stress tests:
> sdk\eventhubs\azure-messaging-eventhubs\test\eventhubs-stress-test\src\CMakeLists.txt
> sdk\tables\azure-data-tables\test\stress\CMakeLists.txt
> sdk\eventhubs\azure-messaging-eventhubs\test\eventhubs-stress-test\CMakeLists.txt

Maybe the core amqp samples need it:
> sdk\core\azure-core-amqp\samples\internal\eventhub_async_writer_sample\CMakeLists.txt
> sdk\core\azure-core-amqp\samples\internal\local_server_sample\CMakeLists.txt

Btw `create_per_service_target_build` was just an example solution. It can be a new macro, if we really need it.

Reply:
I don't see how `new_macro(azure-security-keyvault-administration-test _azure_BUILDING_TESTS)` is substantially different from `target_compile_definitions(azure-security-keyvault-administration-test PRIVATE _azure_BUILDING_TESTS)`.

Reply:
The new (or existing macro) won't need to have `_azure_BUILDING_TESTS` be passed in. That can be inferred by path or target name, by convention.

Reply:
That same logic would have to exist somewhere in codegen, I don't see why it would be buggy. What's the bug?

I still think we can use the  existing macro and not touch 100+ files like this (aside from the fact, we probably don't need to modify test/samples files to begin with).

I get that it is clear and explicit, especially for someone doing it across the board for existing code (like yourself).
The main problem with explicitly setting this at all places, is what happens when we or partner teams don't? Consistency is an answer to reduce the problem of new files/modifications accidentally missing this.

Reply:
Also: How do we plan to codegen this, when we need to?

cc @RickWinter

Reply:
Should codegen end up generating cmakefiles, and samples, it would know perfectly well whether it generates SDK code or samples, so I do not see the problem there.
Should we introduce the macro which does essentially the same thing, we still need to modify the same 100+ files, there is no way around it. If we only add it in the root cmakefile, then we will only get secial values set when the SDK is built from the root, and not from individual directories, like vcpkg does. With my change, we don't get `Cpp2014` from our own code, no matter which make file you used to build it - we simply set the constant on the next line and in the same file where the target is defined. And that is how we will know when we are getting `Cpp2014`, it comes from customer code, and not from building our SDK, and therefore 2014 won't be over-represented in our telemetry - we will simply ignore the values of <=0.